java/source/infodynamics/measures/mixed/kraskov/MutualInfoCalculatorMultiVariateWithDiscreteKraskov.java

```diff
@@ -173,69 +327,129 @@ public void setObservations(double[][] source, double[][] destination, boolean[]
 		throw new RuntimeException("Not implemented yet");
 	}
 
+  /**
+   * Prepare the calculator to receive observations.
+   *
+   * Must be called between {@link #initialise()} and any addObservations
+   * method. Does not need to be called if
+   * {@link setObservations(double[][], int[])} is used.
+   */
 	public void startAddObservations() {
-		throw new RuntimeException("Not implemented yet");
+		vectorOfContinuousObservations = new Vector<double[][]>();
+		vectorOfDiscreteObservations   = new Vector<int[]>();
 	}
 
-	public void finaliseAddObservations() {
-		throw new RuntimeException("Not implemented yet");
-	}
-
-	public void setObservations(double[][] continuousObservations,
-			int[] discreteObservations) throws Exception {
-		if (continuousObservations.length != discreteObservations.length) {
-			throw new Exception("Time steps for observations2 " +
-					discreteObservations.length + " does not match the length " +
-					"of observations1 " + continuousObservations.length);
+  /**
+   * Prepare observations previously added through any addObservations method
+   * for calculations.
+   *
+   * Must be called between any addObservations method and any compute method.
+   * Does not need to be called if {@link setObservations(double[][], int[])}
+   * is used.
+   */
+	public void finaliseAddObservations() throws Exception {
+		if (vectorOfContinuousObservations.size() < 1) {
+			throw new Exception("Cannot compute MI with a null set of data");
 		}
-		if (continuousObservations[0].length == 0) {
-			throw new Exception("Computing MI with a null set of data");
+
+		// First work out the size to allocate the joint vectors, and do the allocation:
+		totalObservations = 0;
+		for (double[][] destination : vectorOfContinuousObservations) {
+			totalObservations += destination.length - timeDiff;
 		}
-		if (continuousObservations[0].length != dimensions) {
-			throw new Exception("The continuous observations do not have the expected number of variables (" + dimensions + ")");
+		continuousData = new double[totalObservations][dimensions];
+		discreteData   = new int[totalObservations];
+		
+		// Construct the joint vectors from the given observations
+		//  (removing redundant data which is outside any timeDiff)
+		int startObservation = 0;
+		Iterator<double[][]> iterator = vectorOfContinuousObservations.iterator();
+		for (int[] dct : vectorOfDiscreteObservations) {
+			double[][] cnt = iterator.next();
+			// Copy the data from these given observations into our master 
+			//  array, aligning them incorporating the timeDiff:
+			MatrixUtils.arrayCopy(cnt, 0, 0,

```

Comment:
At this point, I think we'd need an if statement checking if time_diff < 0 and then pull the data out accommodating that.